// omnes/legacy/packer.js
'use strict';

var onezip = require('onezip/legacy');
var mellow = require('mellow');

module.exports = function (socket, options) {
    if (!options) options = {};

    listen(socket, options);
};

function getRoot(root) {
    if (typeof root === 'function') return root();

    return root;
}

function isRootWin32(path, root) {
    var isRoot = path === '/';
    var isWin32 = process.platform === 'win32';
    var isConfig = root === '/';

    return isWin32 && isRoot && isConfig;
}

function getWin32RootMsg() {
    return Error('Could not pack from/to root on windows!');
}

function check(authCheck) {
    if (authCheck && typeof authCheck !== 'function') throw Error('authCheck should be function!');
}

function listen(socket, options) {
    var authCheck = options.authCheck;
    var prefix = options.prefix || 'salam';
    var root = options.root || '/';

    check(authCheck);

    socket.of(prefix).on('connection', function (socket) {
        if (!authCheck) return connection(root, socket);

        authCheck(socket, function () {
            connection(root, socket);
        });
    });
}

function connection(root, socket) {
    socket.on('pack', function (from, to, files) {
        preprocess('pack', root, socket, from, to, files);
    });

    socket.on('extract', function (from, to) {
        preprocess('extract', root, socket, from, to);
    });
}

function preprocess(op, root, socket, from, to, files) {
    var value = getRoot(root);

    from = mellow.pathToWin(from, value);
    to = mellow.pathToWin(to, value);

    var isRoot = function isRoot(item) {
        return isRootWin32(item, value);
    };

    if (![from, to].some(isRoot)) {
        operate(socket, op, from, to, files);
    } else {
        socket.emit('err', getWin32RootMsg());
        socket.emit('end');
    }
}

function getOperation(op) {
    if (op === 'pack') return onezip.pack;

    if (op === 'extract') return onezip.extract;

    throw Error('op could be pack/extract only!');
}

function operate(socket, op, from, to, files) {
    var fn = getOperation(op);
    var packer = fn(from, to, files);

    packer.on('file', function (name) {
        socket.emit('file', name);
    });

    packer.on('progress', function (percent) {
        socket.emit('progress', percent);
    });

    packer.on('error', function (error) {
        var message = error.message;
        var onAbort = function onAbort() {
            packer.abort();
            socket.removeListener('abort', onAbort);
        };

        socket.emit('err', message);
        socket.on('abort', onAbort);
    });

    packer.on('end', function () {
        socket.emit('end');
    });
}
